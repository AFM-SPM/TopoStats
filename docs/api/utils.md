# Utils Modules

::: topostats.utils
handler: python
options:
docstring_style: numpy
rendering:
show_signature_annotations: true
