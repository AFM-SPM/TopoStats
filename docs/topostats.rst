topostats package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   topostats.logs
   topostats.tracing

Submodules
----------

.. toctree::
   :maxdepth: 4

   topostats.entry_point
   topostats.filters
   topostats.grains
   topostats.grainstats
   topostats.io
   topostats.plotting
   topostats.plottingfuncs
   topostats.processing
   topostats.run_topostats
   topostats.scars
   topostats.statistics
   topostats.theme
   topostats.thresholds
   topostats.utils
   topostats.validation

Module contents
---------------

.. automodule:: topostats
   :members:
   :undoc-members:
   :show-inheritance:
