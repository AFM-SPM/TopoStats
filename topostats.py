"""
File for tracking hard-coded variables (for now)
"""

import argparse

### List of hard-coded variables in pywgwytracing.py

# global variables
gwy_colobar_ln48 = "/module/pixmap/ztype"
gwy_linematchmethod_ln51= '/module/linematch/method'
gwy_linematchmaxdegree_ln52 = "/module/linematch/max_degree"
gwy_polylevel_coldegree_ln53 = "/module/polylevel/col_degree"

# heightediting()
datafield_grains_mark_height_ln128 = int()
gwy_polylevel_masking_ln131 = int()
gwy_linematch_masking_ln135 = int()
filterwidth_ln144 = float()

# editfile()
datafield_markheight_ln170 = int()
gwy_polylevel_masking_ln173 = int()
gwy_linematch_masking_ln177 = gwy_linematch_masking_ln135
gwy_polylevel_masking_ln181 = int()
datafield_gaussfilter_ln195 = int()

# grainfinding()
grain_gauss_ln211 = float()
datafield_mask_outliers_ln220 = float()

# remove_large_objects()
datafield_mask_outliers_ln257 = float()

# remove_small_objects()
datafield_mask_outliers_ln283 = float()

# grain_thinning()
gaussian_size_ln488 = float()

# savefiles()
savefile_type_ln535 = ["/module/pixmap/xytype"]

# saveunknownfiles()
savefile_type_ln573 = ["/module/pixmap/xytype"]

# savecroppedfiles()
savefile_type_ln645 = ["/module/pixmap/xytype"]

# __main__()
path_ln709 = str()
fileend_ln712 = []
filetype_ln713 = str()
extension_ln715 = str()
minheightscale_ln717 = float()
maxheightscale_ln718 = float()
minarea_ln722 = float()
maxdeviation_ln726 = float()
mindeviation_ln727 = float()
cropwidth_ln733 = float()
splitwidth_ln734 = float()
bins_ln736 = int()
channels_ln755 = list()


### dnatracing.py

# getNumpyArraysfromGwyddion()
sigma_ln94 = float()

# getDisorderedTrace()
sigma_ln109 = float()

# purgeObviousCrap()
maxlen_ln126 = int()

# getFittedTraces()
index_width_ln212 = int()

# getSplinedTraces()
step_size_ln376 = float()
interp_size_ln377 = float()

### tracingfuncs.py

# I didn't find any hard-coded variables so far

### statsplotting.py
### TFOplot.py
### traceplotting.py

# not checking these for now due to ongoing merging project
