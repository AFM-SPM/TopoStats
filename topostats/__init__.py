"""Topostats"""
from importlib.metadata import version

from .logs.logs import setup_logger

LOGGER = setup_logger()

release = version("topostats")
__version__ = ".".join(release.split("."[:2]))
