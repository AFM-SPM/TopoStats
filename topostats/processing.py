"""Functions for procesing data."""
from collections import defaultdict
from pathlib import Path
from typing import Dict, Union, List

import numpy as np
import pandas as pd

from topostats import __version__
from topostats.filters import Filters
from topostats.grains import Grains
from topostats.grainstats import GrainStats
from topostats.io import get_out_path, save_array
from topostats.logs.logs import setup_logger, LOGGER_NAME
from topostats.plottingfuncs import Images
from topostats.tracing.dnatracing import dnaTrace, traceStats
from topostats.utils import create_empty_dataframe

# pylint: disable=broad-except
# pylint: disable=line-too-long
# pylint: disable=too-many-arguments
# pylint: disable=too-many-branches
# pylint: disable=too-many-locals
# pylint: disable=too-many-statements
# pylint: disable=too-many-nested-blocks
# pylint: disable=unnecessary-dict-index-lookup

LOGGER = setup_logger(LOGGER_NAME)


def process_scan(
    img_path_px2nm: Dict[str, Union[np.ndarray, Path, float]],
    base_dir: Union[str, Path],
    filter_config: dict,
    grains_config: dict,
    grainstats_config: dict,
    dnatracing_config: dict,
    plotting_config: dict,
    output_dir: Union[str, Path] = "output",
) -> None:
    """Process a single image, filtering, finding grains and calculating their statistics.

    Parameters
    ----------
    img_path_px2nm : Dict[str, Union[np.ndarray, Path, float]]
        A dictionary with keys 'image', 'img_path' and 'px_2_nm' containing a file or frames' image, it's path and it's pixel to namometre scaling value.
    base_dir : Union[str, Path]
        Directory to recursively search for files, if not specified the current directory is scanned.
    filter_config : dict
        Dictionary of configuration options for running the Filter stage.
    grains_config : dict
        Dictionary of configuration options for running the Grain detection stage.
    grainstats_config : dict
        Dictionary of configuration options for running the Grain Statistics stage.
    dnatracing_config : dict
        Dictionary of configuration options for running the DNA Tracing stage.
    plotting_config : dict
        Dictionary of configuration options for plotting figures.
    output_dir : Union[str, Path]
        Directory to save output to, it will be created if it does not exist. If it already exists then it is possible
        that output will be over-written.


    Results
    -------
    None

    Results are written to CSV and images produced in configuration options request them.
    """

    image = img_path_px2nm["image"]
    image_path = img_path_px2nm["img_path"]
    pixel_to_nm_scaling = img_path_px2nm["px_2_nm"]
    filename = image_path.name

    LOGGER.info(f"Processing : {filename}")
    core_out_path = get_out_path(image_path, base_dir, output_dir).parent / "processed"
    core_out_path.mkdir(parents=True, exist_ok=True)
    filter_out_path = core_out_path / filename / "filters"
    grain_out_path = core_out_path / filename / "grains"
    if plotting_config["image_set"] == "all":
        filter_out_path.mkdir(exist_ok=True, parents=True)
        Path.mkdir(grain_out_path / "above", parents=True, exist_ok=True)
        Path.mkdir(grain_out_path / "below", parents=True, exist_ok=True)

    # Filter Image
    if filter_config["run"]:
        filter_config.pop("run")
        LOGGER.info(f"[{filename}] Image dimensions: {image.shape}")
        LOGGER.info(f"[{filename}] : *** Filtering ***")
        filtered_image = Filters(
            image,
            filename,
            pixel_to_nm_scaling,
            **filter_config,
        )
        filtered_image.filter_image()

        # Optionally plot filter stage
        if plotting_config["run"]:
            plotting_config.pop("run")
            LOGGER.info(f"[{filename}] : Plotting Filtering Images")
            # Update PLOT_DICT with pixel_to_nm_scaling (can't add _output_dir since it changes)
            plot_opts = {"pixel_to_nm_scaling": pixel_to_nm_scaling}
            for image, options in plotting_config["plot_dict"].items():
                plotting_config["plot_dict"][image] = {**options, **plot_opts}
            # Generate plots
            for plot_name, array in filtered_image.images.items():
                if plot_name not in ["scan_raw"]:
                    if plot_name == "extracted_channel":
                        array = np.flipud(array.pixels)
                    plotting_config["plot_dict"][plot_name]["output_dir"] = filter_out_path
                    try:
                        Images(array, **plotting_config["plot_dict"][plot_name]).plot_and_save()
                        Images(array, **plotting_config["plot_dict"][plot_name]).plot_histogram_and_save()
                    except AttributeError:
                        LOGGER.info(f"[{filename}] Unable to generate plot : {plot_name}")
            plotting_config["run"] = True
        # Always want the 'z_threshed' plot (aka "Height Thresholded") but in the core_out_path
        plot_name = "z_threshed"
        plotting_config["plot_dict"][plot_name]["output_dir"] = core_out_path
        Images(
            filtered_image.images["gaussian_filtered"],
            filename=filename,
            **plotting_config["plot_dict"][plot_name],
        ).plot_and_save()
        # Save the z_threshed image (aka "Height_Thresholded") numpy array
        save_array(
            array=filtered_image.images["gaussian_filtered"],
            outpath=core_out_path,
            filename=filename,
            array_type="height_thresholded",
        )

    else:
        LOGGER.error(
            "You have not included running the initial filter stage. This is required for all subsequent "
            "stages of processing. Please check your configuration file."
        )

    # Find Grains :
    if grains_config["run"]:
        grains_config.pop("run")
        try:
            LOGGER.info(f"[{filename}] : *** Grain Finding ***")
            grains = Grains(
                image=filtered_image.images["gaussian_filtered"],
                filename=filename,
                pixel_to_nm_scaling=pixel_to_nm_scaling,
                **grains_config,
            )
            grains.find_grains()
            for direction, _ in grains.directions.items():
                LOGGER.info(
                    f"[{filename}] : Grains found for direction {direction} : {len(grains.region_properties[direction])}"
                )
                if len(grains.region_properties[direction]) == 0:
                    LOGGER.warning(f"[{filename}] : No grains found for direction {direction}")
        except Exception as e:
            LOGGER.error(f"[{filename}] : An error occured during grain finding, skipping grainstats and dnatracing.")
            LOGGER.error(f"[{filename}] : The error: {e}")
            results = create_empty_dataframe()
        else:
            for direction, region_props in grains.region_properties.items():
                if len(region_props) == 0:
                    LOGGER.warning(f"[{filename}] : No grains found for the {direction} direction.")
            # Optionally plot grain finding stage if we have found grains and plotting is required
            if plotting_config["run"]:
                plotting_config.pop("run")
                LOGGER.info(f"[{filename}] : Plotting Grain Finding Images")
                for direction, image_arrays in grains.directions.items():
                    LOGGER.info(f"[{filename}] : Plotting {direction} Grain Finding Images")
                    for plot_name, array in image_arrays.items():
                        LOGGER.info(f"[{filename}] : Plotting {plot_name} image")
                        plotting_config["plot_dict"][plot_name]["output_dir"] = grain_out_path / f"{direction}"
                        Images(array, **plotting_config["plot_dict"][plot_name]).plot_and_save()
                    # Make a plot of coloured regions with bounding boxes
                    plotting_config["plot_dict"]["bounding_boxes"]["output_dir"] = grain_out_path / f"{direction}"
                    Images(
                        grains.directions[direction]["coloured_regions"],
                        **plotting_config["plot_dict"]["bounding_boxes"],
                        region_properties=grains.region_properties[direction],
                    ).plot_and_save()
                    plotting_config["plot_dict"]["coloured_boxes"]["output_dir"] = grain_out_path / f"{direction}"
                    Images(
                        grains.directions[direction]["labelled_regions_02"],
                        **plotting_config["plot_dict"]["coloured_boxes"],
                        region_properties=grains.region_properties[direction],
                    ).plot_and_save()
                    # Always want mask_overlay (aka "Height Thresholded with Mask") but in core_out_path
                    plot_name = "mask_overlay"
                    plotting_config["plot_dict"][plot_name]["output_dir"] = core_out_path
                    Images(
                        filtered_image.images["gaussian_filtered"],
                        filename=f"{filename}_{direction}_masked",
                        masked_array=grains.directions[direction]["removed_small_objects"],
                        **plotting_config["plot_dict"][plot_name],
                    ).plot_and_save()

                plotting_config["run"] = True
            else:
                LOGGER.info(f"[{filename}] : Plotting disabled for Grain Finding Images")

            # Grainstats :
            # Calculate statistics if required
            if grainstats_config["run"]:
                grainstats_config.pop("run")
                # Grain Statistics :
                try:
                    LOGGER.info(f"[{filename}] : *** Grain Statistics ***")
                    grain_plot_dict = {
                        key: value
                        for key, value in plotting_config["plot_dict"].items()
                        if key in ["grain_image", "grain_mask", "grain_mask_image"]
                    }
                    grainstats = {}
                    # There are two layers to process those above the given threshold and those below
                    for direction, _ in grains.directions.items():
                        if len(grains.region_properties[direction]) == 0:
                            LOGGER.warning(
                                f"[{filename}] : No grains exist for the {direction} direction. Skipping grainstats and DNAtracing."
                            )
                            grainstats[direction] = create_empty_dataframe()
                        else:
                            grainstats[direction], grains_plot_data = GrainStats(
                                data=filtered_image.images["gaussian_filtered"],
                                labelled_data=grains.directions[direction]["labelled_regions_02"],
                                pixel_to_nanometre_scaling=pixel_to_nm_scaling,
                                direction=direction,
                                base_output_dir=grain_out_path,
                                image_name=filename,
                                plot_opts=grain_plot_dict,
                                **grainstats_config,
                            ).calculate_stats()
                            grainstats[direction]["threshold"] = direction
                            # Plot grains
                            if plotting_config["image_set"] == "all":
                                LOGGER.info(f"[{filename}] : Plotting grain images for direction: {direction}.")
                                for plot_data in grains_plot_data:
                                    LOGGER.info(
                                        f"[{filename}] : Plotting grain image {plot_data['filename']} for direction: {direction}."
                                    )
                                    Images(
                                        data=plot_data["data"],
                                        output_dir=plot_data["output_dir"],
                                        filename=plot_data["filename"],
                                        **plotting_config["plot_dict"][plot_data["name"]],
                                    ).plot_and_save()
                    # Set tracing_stats_df in light of direction
                    if grains_config["direction"] == "both":
                        grainstats_df = pd.concat([grainstats["below"], grainstats["above"]])
                    elif grains_config["direction"] == "above":
                        grainstats_df = grainstats["above"]
                    elif grains_config["direction"] == "below":
                        grainstats_df = grainstats["below"]
                except Exception:
                    LOGGER.info(
                        f"[{filename}] : Errors occurred whilst calculating grain statistics. Skipping DNAtracing."
                    )
                    results = create_empty_dataframe()
                else:
                    # Run dnatracing
                    try:
                        if dnatracing_config["run"]:
                            dnatracing_config.pop("run")
                            LOGGER.info(f"[{filename}] : *** DNA Tracing ***")
                            dna_traces = defaultdict()
                            tracing_stats = defaultdict()
                            for direction, _ in grainstats.items():
                                dna_traces[direction] = dnaTrace(
                                    full_image_data=filtered_image.images["gaussian_filtered"].T,
                                    grains=grains.directions[direction]["labelled_regions_02"],
                                    filename=filename,
                                    pixel_size=pixel_to_nm_scaling,
                                    **dnatracing_config,
                                )
                                dna_traces[direction].trace_dna()
                                tracing_stats[direction] = traceStats(
                                    trace_object=dna_traces[direction], image_path=image_path
                                )
                                tracing_stats[direction].df["threshold"] = direction
                            # Set tracing_stats_df in light of direction
                            if grains_config["direction"] == "both":
                                tracing_stats_df = pd.concat([tracing_stats["below"].df, tracing_stats["above"].df])
                            elif grains_config["direction"] == "above":
                                tracing_stats_df = tracing_stats["above"].df
                            elif grains_config["direction"] == "below":
                                tracing_stats_df = tracing_stats["below"].df
                            LOGGER.info(
                                f"[{filename}] : Combining {direction} grain statistics and dnatracing statistics"
                            )
                            # NB - Merge on image, molecule and threshold because we may have above and below molecueles which
                            #      gives duplicate molecule numbers as they are processed separately, if tracing stats
                            #      are not available (because skeleton was too small), grainstats are still retained.
                            results = grainstats_df.merge(
                                tracing_stats_df, on=["image", "threshold", "molecule_number"], how="left"
                            )
                            results["basename"] = image_path.parent
                        else:
                            LOGGER.info(
                                f"[{filename}] Calculation of DNA Tracing disabled, returning grainstats data frame."
                            )
                            results = grainstats_df
                            results["basename"] = image_path.parent
                    except Exception:
                        # If no results we need a dummy dataframe to return.
                        LOGGER.warning(
                            f"[{filename}] : Errors occurred whilst calculating DNA tracing statistics, "
                            "returning grain statistics"
                        )
                        results = grainstats_df
                        results["basename"] = image_path.parent
            else:
                LOGGER.info(f"[{filename}] Calculation of grainstats disabled, returning empty data frame.")
                results = create_empty_dataframe()
    else:
        LOGGER.info(f"[{filename}] Detection of grains disabled, returning empty data frame.")
        results = create_empty_dataframe()

    return image_path, results


def check_run_steps(filter_run: bool, grains_run: bool, grainstats_run: bool, dnatracing_run: bool) -> None:
    """Check options for running steps (Filter, Grain, Grainstats and DNA tracing) are logically consistent.

    This checks that earlier steps required are enabled.

    Parameters
    ----------
    filter_run: bool
        Flag for running Filtering.
    grains_run: bool
        Flag for running Grains.
    grainstats_run: bool
        Flag for running GrainStats.
    dnatracing_run: bool
        Flag for running DNA Tracing.

    Returns
    -------
    None
    """
    if dnatracing_run:
        if grainstats_run is False:
            LOGGER.error("DNA tracing enabled but Grainstats disabled. Please check your configuration file.")
        elif grains_run is False:
            LOGGER.error("DNA tracing enabled but Grains disabled. Please check your configuration file.")
        elif filter_run is False:
            LOGGER.error("DNA tracing enabled but Filters disabled. Please check your configuration file.")
        else:
            LOGGER.info("Configuration run options are consistent, processing can proceed.")
    elif grainstats_run:
        if grains_run is False:
            LOGGER.error("Grainstats enabled but Grains disabled. Please check your configuration file.")
        elif filter_run is False:
            LOGGER.error("Grainstats enabled but Filters disabled. Please check your configuration file.")
        else:
            LOGGER.info("Configuration run options are consistent, processing can proceed.")
    elif grains_run:
        if filter_run is False:
            LOGGER.error("Grains enabled but Filters disabled. Please check your configuration file.")
        else:
            LOGGER.info("Configuration run options are consistent, processing can proceed.")
    else:
        LOGGER.info("Configuration run options are consistent, processing can proceed.")


def completion_message(config: Dict, img_files: List, summary_config: Dict, images_processed: int) -> None:
    """Print a completion message summarising images processed.

    Parameters
    ----------
    config: dict
        Configuration dictionary.
    img_files: list()
        List of found image paths.
    summary_config: dict(
        Configuration for plotting summary statistics.
    images_processed: int
        Pandas DataFrame of results.

    Results
    -------
    None
    """

    if summary_config is not None:
        distribution_plots_message = str(summary_config["output_dir"])
    else:
        distribution_plots_message = "Disabled. Enable in config 'summary_stats/run' if needed."
    LOGGER.info(
        f"\n\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ COMPLETE ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\n"
        f"  TopoStats Version           : {__version__}\n"
        f"  Base Directory              : {config['base_dir']}\n"
        f"  File Extension              : {config['file_ext']}\n"
        f"  Files Found                 : {len(img_files)}\n"
        f"  Successfully Processed^1    : {images_processed} ({(images_processed * 100) / len(img_files)}%)\n"
        f"  Configuration               : {config['output_dir']}/config.yaml\n"
        f"  All statistics              : {str(config['output_dir'])}/all_statistics.csv\n"
        f"  Distribution Plots          : {distribution_plots_message}\n\n"
        f"  Email                       : topostats@sheffield.ac.uk\n"
        f"  Documentation               : https://afm-spm.github.io/topostats/\n"
        f"  Source Code                 : https://github.com/AFM-SPM/TopoStats/\n"
        f"  Bug Reports/Feature Request : https://github.com/AFM-SPM/TopoStats/issues/new/choose\n"
        f"  Citation File Format        : https://github.com/AFM-SPM/TopoStats/blob/main/CITATION.cff\n\n"
        f"  ^1 Successful processing of an image is detection of grains and calculation of at least\n"
        f"     grain statistics. If these have been disabled the percentage will be 0.\n\n"
        f"  If you encounter bugs/issues or have feature requests please report them at the above URL\n"
        f"  or email us.\n\n"
        f"  If you have found TopoStats useful please consider citing it. A Citation File Format is\n"
        f"  linked above and available from the Source Code page.\n"
        f"~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\n"
    )
