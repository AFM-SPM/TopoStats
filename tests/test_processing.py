"""Test end-to-end running of topostats."""
from pathlib import Path

# pylint: disable=deprecated-module
import imghdr
import numpy as np
import pytest

from topostats.io import LoadScans
from topostats.processing import check_run_steps, process_scan
from topostats.utils import update_plotting_config

BASE_DIR = Path.cwd()


# Can't see a way of paramterising with pytest-regtest as it writes to a file based on the file/function
# so instead we run three regression tests.
def test_process_scan_below(regtest, tmp_path, process_scan_config: dict, load_scan_data: LoadScans) -> None:
    """Regression test for checking the process_scan functions correctly"""
    process_scan_config["grains"]["direction"] = "below"
    img_dic = load_scan_data.img_dict
    _, results = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    # Remove the basename column as this differs on CI
    results.drop(["basename"], axis=1, inplace=True)
    print(results.to_string(), file=regtest)  # noqa: T201


def test_process_scan_above(regtest, tmp_path, process_scan_config: dict, load_scan_data: LoadScans) -> None:
    """Regression test for checking the process_scan functions correctly"""
    img_dic = load_scan_data.img_dict
    _, results = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    # Remove the Basename column as this differs on CI
    results.drop(["basename"], axis=1, inplace=True)
    print(results.to_string(), file=regtest)  # noqa: T201


def test_process_scan_both(regtest, tmp_path, process_scan_config: dict, load_scan_data: LoadScans) -> None:
    """Regression test for checking the process_scan functions correctly"""
    process_scan_config["grains"]["direction"] = "both"
    img_dic = load_scan_data.img_dict
    _, results = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    # Remove the Basename column as this differs on CI
    results.drop(["basename"], axis=1, inplace=True)
    print(results.to_string(), file=regtest)  # noqa: T201


@pytest.mark.parametrize(
    "image_set, expected",
    [
        ("core", False),
        ("all", True),
    ],
)
def test_save_cropped_grains(
    tmp_path: Path, process_scan_config: dict, load_scan_data: LoadScans, image_set, expected
) -> None:
    """Tests if cropped grains are saved only when image set is 'all' rather than 'core'."""
    process_scan_config["plotting"]["image_set"] = image_set
    process_scan_config["plotting"] = update_plotting_config(process_scan_config["plotting"])

    img_dic = load_scan_data.img_dict
    _, _ = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )

    assert (
        Path.exists(tmp_path / "tests/resources/processed/minicircle/grains/above" / "minicircle_grain_image_0.png")
        == expected
    )
    assert (
        Path.exists(tmp_path / "tests/resources/processed/minicircle/grains/above" / "minicircle_grain_mask_0.png")
        == expected
    )
    assert (
        Path.exists(
            tmp_path / "tests/resources/processed/minicircle/grains/above" / "minicircle_grain_mask_image_0.png"
        )
        == expected
    )


@pytest.mark.parametrize("extension", [("png"), ("tiff")])
def test_save_format(process_scan_config: dict, load_scan_data: LoadScans, tmp_path: Path, extension: str):
    """Tests if save format applied to cropped images"""
    process_scan_config["plotting"]["image_set"] = "all"
    process_scan_config["plotting"]["save_format"] = extension
    process_scan_config["plotting"] = update_plotting_config(process_scan_config["plotting"])

    img_dic = load_scan_data.img_dict
    _, _ = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )

    assert (
        imghdr.what(
            tmp_path / "tests/resources/processed/minicircle/grains/above" / f"minicircle_grain_image_0.{extension}"
        )
        == extension
    )


@pytest.mark.parametrize(
    "filter_run, grains_run, grainstats_run, dnatracing_run, log_msg",
    [
        (
            False,
            False,
            False,
            True,
            "DNA tracing enabled but Grainstats disabled. Please check your configuration file.",
        ),
        (
            False,
            False,
            True,
            True,
            "DNA tracing enabled but Grains disabled. Please check your configuration file.",
        ),
        (
            False,
            True,
            True,
            True,
            "DNA tracing enabled but Filters disabled. Please check your configuration file.",
        ),
        (
            False,
            False,
            True,
            False,
            "Grainstats enabled but Grains disabled. Please check your configuration file.",
        ),
        (
            False,
            True,
            True,
            False,
            "Grainstats enabled but Filters disabled. Please check your configuration file.",
        ),
        (
            False,
            True,
            False,
            False,
            "Grains enabled but Filters disabled. Please check your configuration file.",
        ),
        (
            True,
            False,
            False,
            False,
            "Configuration run options are consistent, processing can proceed.",
        ),
        (
            True,
            True,
            False,
            False,
            "Configuration run options are consistent, processing can proceed.",
        ),
        (
            True,
            True,
            True,
            False,
            "Configuration run options are consistent, processing can proceed.",
        ),
        (
            True,
            True,
            True,
            True,
            "Configuration run options are consistent, processing can proceed.",
        ),
    ],
)
def test_check_run_steps(
    filter_run: bool,
    grains_run: bool,
    grainstats_run: bool,
    dnatracing_run: bool,
    log_msg: str,
    caplog,
) -> None:
    """Test the logic which checks whether enabled processing options are consistent."""
    check_run_steps(filter_run, grains_run, grainstats_run, dnatracing_run)
    assert log_msg in caplog.text


# noqa: disable=too-many-arguments
# pylint: disable=too-many-arguments
@pytest.mark.parametrize(
    "filter_run, grains_run, grainstats_run, dnatracing_run, log_msg1, log_msg2",
    [
        (
            False,
            False,
            False,
            False,
            "You have not included running the initial filter stage.",
            "Please check your configuration file.",
        ),
        (
            True,
            False,
            False,
            False,
            "Detection of grains disabled, returning empty data frame.",
            "13-gaussian_filtered",
        ),
        (
            True,
            True,
            False,
            False,
            "Calculation of grainstats disabled, returning empty data frame.",
            "22-labelled_image_bboxes",
        ),
        (
            True,
            True,
            True,
            False,
            "Processing grain",
            "Calculation of DNA Tracing disabled, returning grainstats data frame.",
        ),
        (
            True,
            True,
            True,
            True,
            "There are 15 circular and 6 linear DNA molecules found in the image",
            "Combining above grain statistics and dnatracing statistics",
        ),
    ],
)
def test_process_stages(
    process_scan_config: dict,
    load_scan_data: LoadScans,
    tmp_path: Path,
    filter_run: bool,
    grains_run: bool,
    grainstats_run: bool,
    dnatracing_run: bool,
    log_msg1: str,
    log_msg2: str,
    caplog,
) -> None:
    """Regression test for checking the process_scan functions correctly when specific sections are disabled.

    Currently there is no test for having later stages (e.g. DNA Tracing or Grainstats) enabled when Filters and/or
    Grainstats are disabled. Whislt possible it is expected that users understand the need to run earlier stages before
    later staged can run and do not disable earlier stages."""
    img_dic = load_scan_data.img_dict
    process_scan_config["filter"]["run"] = filter_run
    process_scan_config["grains"]["run"] = grains_run
    process_scan_config["grainstats"]["run"] = grainstats_run
    process_scan_config["dnatracing"]["run"] = dnatracing_run
    _, _ = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )

    assert log_msg1 in caplog.text
    assert log_msg2 in caplog.text


def test_process_scan_no_grains(process_scan_config: dict, load_scan_data: LoadScans, tmp_path: Path, caplog) -> None:
    """Test handling no grains found during grains.find_grains()."""
    img_dic = load_scan_data.img_dict
    process_scan_config["grains"]["threshold_std_dev"]["above"] = 1000
    process_scan_config["filter"]["remove_scars"]["run"] = False
    _, _ = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    assert "Grains found for direction above : 0" in caplog.text
    assert "There are 0 circular and 0 linear DNA molecules found in the image" in caplog.text
    assert "No grains exist for the above direction. Skipping grainstats and DNAtracing." in caplog.text


def test_process_scan_align_grainstats_dnatracing(
    process_scan_config: dict, load_scan_data: LoadScans, tmp_path: Path
) -> None:
    """Test that molecule numbers from dnatracing align with those from grainstats when some grains are removed from
    tracing because they are too small.

    By setting processing parameters as below two molecules are pruged for being too small after skeletonisation and so
    do not have DNA tracing statistics (but they do have Grain Statistics)."""
    img_dic = load_scan_data.img_dict
    process_scan_config["filter"]["remove_scars"]["run"] = False
    process_scan_config["grains"]["absolute_area_threshold"]["above"] = [150, 3000]
    process_scan_config["dnatracing"]["min_skeleton_size"] = 10
    _, results = process_scan(
        img_path_px2nm=img_dic["minicircle"],
        base_dir=BASE_DIR,
        filter_config=process_scan_config["filter"],
        grains_config=process_scan_config["grains"],
        grainstats_config=process_scan_config["grainstats"],
        dnatracing_config=process_scan_config["dnatracing"],
        plotting_config=process_scan_config["plotting"],
        output_dir=tmp_path,
    )
    tracing_to_check = ["contour_lengths", "circular", "end_to_end_distance"]
    assert results.shape == (24, 25)
    assert np.isnan(results.loc[8, "contour_lengths"])
    assert np.isnan(sum(results.loc[8, tracing_to_check]))
    assert np.isnan(results.loc[22, "contour_lengths"])
    assert np.isnan(sum(results.loc[22, tracing_to_check]))
