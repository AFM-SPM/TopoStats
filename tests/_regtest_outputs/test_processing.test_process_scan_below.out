                     centre_x      centre_y    radius_min    radius_max   radius_mean  radius_median    height_min    height_max  height_median   height_mean        volume          area  area_cartesian_bbox  smallest_bounding_width  smallest_bounding_length  smallest_bounding_area  aspect_ratio threshold     max_feret     min_feret       image  contour_lengths  circular  end_to_end_distance
molecule_number                                                                                                                                                                                                                                                                                                                                                                                          
0                3.124616e-07  1.354863e-07  6.688903e-10  7.153789e-09  4.011343e-09   4.131417e-09 -8.118862e-10 -4.370350e-10  -5.745181e-10 -5.807966e-10 -2.919781e-26  5.027201e-17         1.061569e-16             5.569782e-09              1.364449e-08            7.599684e-17      2.449735     below  1.391125e-08  5.569782e-09  minicircle     6.548362e-08     False         5.969060e-09
